gateway/blocks_gateway.go

```diff
@@ -0,0 +1,484 @@
+	}
+
+	rootCodec := nd.Cid().Prefix().GetCodec()
+	// This covers both Raw blocks and terminal IPLD codecs like dag-cbor and dag-json
+	// Note: while only cbor, json, dag-cbor, and dag-json are currently supported by gateways this could change
+	if rootCodec != uint64(mc.DagPb) {
+		return md, files.NewBytesFile(nd.RawData()), nil
+	}
+
+	// This code path covers full graph, single file/directory, and range requests
+	f, err := ufile.NewUnixfsFile(ctx, api.dagService, nd)
+	// Note: there is an assumption here that non-UnixFS dag-pb should not be returned which is currently valid
+	if err != nil {
+		return md, nil, err
+	}
+	return md, f, nil
+}
+
+func (api *BlocksGateway) GetRange(ctx context.Context, path ImmutablePath, ranges ...GetRange) (ContentPathMetadata, files.File, error) {
+	roots, lastSeg, err := api.getPathRoots(ctx, path)
+	if err != nil {
+		return ContentPathMetadata{}, nil, err
+	}
+
+	md := ContentPathMetadata{
+		PathSegmentRoots: roots,
+		LastSegment:      lastSeg,
+	}
+
+	lastRoot := lastSeg.Cid()
+
+	nd, err := api.dagService.Get(ctx, lastRoot)
+	if err != nil {
+		return ContentPathMetadata{}, nil, err
+	}
+
+	// This code path covers full graph, single file/directory, and range requests
+	n, err := ufile.NewUnixfsFile(ctx, api.dagService, nd)
+	if err != nil {
+		return md, nil, err
+	}
+	f, ok := n.(files.File)
+	if !ok {
+		// TODO: Should we use either a general IPLD error code here, or one for gateways?
+		return ContentPathMetadata{}, nil, NewErrorResponse(fmt.Errorf("can only do range requests on files, but did not get a file"), http.StatusBadRequest)
+	}
+
+	return md, f, nil
+}
+
+func (api *BlocksGateway) GetAll(ctx context.Context, path ImmutablePath) (ContentPathMetadata, files.Node, error) {
+	roots, lastSeg, err := api.getPathRoots(ctx, path)
+	if err != nil {
+		return ContentPathMetadata{}, nil, err
+	}
+
+	md := ContentPathMetadata{
+		PathSegmentRoots: roots,
+		LastSegment:      lastSeg,
+	}
+
+	lastRoot := lastSeg.Cid()
+
+	nd, err := api.dagService.Get(ctx, lastRoot)
+	if err != nil {
+		return ContentPathMetadata{}, nil, err
+	}
+
+	// This code path covers full graph, single file/directory, and range requests
+	n, err := ufile.NewUnixfsFile(ctx, api.dagService, nd)
+	if err != nil {
+		return md, nil, err
+	}
+	return md, n, nil
+}
+
+func (api *BlocksGateway) GetBlock(ctx context.Context, path ImmutablePath) (ContentPathMetadata, files.File, error) {
+	roots, lastSeg, err := api.getPathRoots(ctx, path)
+	if err != nil {
+		return ContentPathMetadata{}, nil, err
+	}
+
+	md := ContentPathMetadata{
+		PathSegmentRoots: roots,
+		LastSegment:      lastSeg,
+	}
+
+	lastRoot := lastSeg.Cid()
+
+	nd, err := api.dagService.Get(ctx, lastRoot)
+	if err != nil {
+		return ContentPathMetadata{}, nil, err
+	}
+
+	return md, files.NewBytesFile(nd.RawData()), nil
+}
+
+func (api *BlocksGateway) Head(ctx context.Context, path ImmutablePath) (ContentPathMetadata, files.Node, error) {
```

Comment:
nit: would this be better for consistency (with `GetAll` etc)?
```suggestion
func (api *BlocksGateway) GetHead(ctx context.Context, path ImmutablePath) (ContentPathMetadata, files.Node, error) {
```

Reply:
I think the idea here was that the methods were coupled with the HTTP method? If we rename to `GetHead` there will no longer be a distinction.